pyproject.toml

```diff
@@ -124,7 +125,7 @@ omit = [
 ]
 
 [tool.ruff]
-target-version = "py311"
+target-version = "py39"
```

Comment:
oh sorry ya, this is because ruff uses this as the minimum target, so it was constantly suggesting changes to the code that were incompatible with our lowest version. (specifically `Union[X,Y]` to `X | Y` which 3.9 doesn't support). 3.9 should be EOL next october, so basically this value should always stay pegged to our lowest version we support and bumped then (rather than the max version, which was 3.11 at the time i set this iirc)